Fix JRuby builds / minimum ruby version

eval opened the pull request:
Recent JRuby builds were failing with errors "Error: Unavailable version 9.2.21.0 for jruby on ubuntu-24.04".

This PR ensures JRuby uses ubuntu-22.04.

It also prunes builds for CRuby 2.5 and 2.6.

eval commented:
As Ruby 2.7 is now [2 years EOL](https://www.ruby-lang.org/en/downloads/branches/), maybe a good moment to prune that version as well, WDYT @radar?

fz-ryanbigg commented:
Yep, trim it. While you're there, also trim 3.0 (EOL'D almost a year ago). Keep 3.1, despite it EOL'ing in 4 days time.

eval committed:
Fix JRuby and ruby-head builds

JRuby needs an older ubuntu.
Ruby >3.4 no longer has benchmark as default gem.

eval committed:
Require ruby 3.1

Ruby 3.0 EOL was 2024-04-23.

**NOTE:** The next release after this is merged should be v2.9.0.

eval commented:
@fz-ryanbigg 3.1 it is 👍🏻

eval commented:
@fz-ryanbigg good to go?
